Add new gall guide. (#1249)
urbit/urbit.org: content/docs/userspace/gall-guide/5-cards.md

> +++

Reviewer: I think it would be useful to include somewhere on this page a diagram of a typically card with labelled parts, like a picture of `[%pass /some/wire %agent [~some-ship %some-agent] %watch-as %some-mark /some/path]` as a rectangular box with dividers between each section, labeling what they are. Of course this info is already implicit and there's a form of this in the diagram at the bottom, but more diagrams are frequently helpful. Something like https://media.urbit.org/docs/arvo/scry-diagram-v2.svg

title = "5. Cards"
weight = 25
template = "doc.html"
+++

As we previously discussed, most arms of an agent core produce a cell of
`[effects new-agent-core]`, and the type we use for this is typically `(quip card _this)`. We've covered `_this`, but we haven't yet looked at `card`s in
detail. That's what we'll do here. In explaining `card`s we'll touch on some
concepts relating to the mechanics of pokes, subscriptions and other things
we've not yet covered. Don't worry if you don't understand how it all fits
together yet, we just want to give you a basic idea of `card`s so we can then

> dig into how they work in practice.

Reviewer 2: This intro should probably explicitly call out cards as "effects".

## `card` type

The `card:agent:gall` type (henceforth just `card`) has a slightly complex
structure, so we'll walk through it step-by-step.

`lull.hoon` defines a `card` like so:

```hoon
+$  card  (wind note gift)
```

A `wind` is defined in `arvo.hoon` as:

```hoon
++  wind
  |$  [a b]

>   $%  [%pass p=path q=a]

Reviewer 2: Huh, yeah, I guess these are `path`s canonically. Wondering if it wouldn't be better to tell a little white lie here and put `wire` here instead, to help ease the explanation about wires below.

`wire`/`path` distinction is merely to differentiate their uses. When the agent
or vane responds to your request, your agent will see the `wire` as well as the
response itself. Your agent can then check the `wire` and maybe do different
things depending on its content. Both the
[`wire`](/docs/userspace/gall-guide/types#wire) and
[`path`](/docs/userspace/gall-guide/types#path) types are are covered in the
[types reference](/docs/userspace/gall-guide/types). We'll show how `wire`s are
practically used later on.

The type of the next field is a `note:agent:gall` (henceforth just `note`), which
`lull.hoon` defines as:

```hoon
+$  note
  $%  [%agent [=ship name=term] =task]
      [%arvo note-arvo]
      [%pyre =tang]
  ==
```

- An `%agent` `note` is a request to another Gall agent, either local or on a
  remote ship. The `ship` and `name` fields are just the target ship and agent
  name. The `task` is the request itself, we'll discuss it separately
  [below](#task).
- An `%arvo` `note` is a request to a vane. We'll discuss such requests
  [below](#note-arvo).
- A `%pyre` `note` is used to abort an event. It's mostly used internally by

>   Kiln, it's unlikely you'd use it in your own agent. The `tang` contains an

Reviewer: ```suggestion
  `kiln` (a submodule of `%hood`), it's unlikely you'd use it in your own agent. The `tang` contains an
```
Dunno if we have any convention for how to stylize submodules of `%hood`, but just writing it like a normal word feels wrong, so I'm suggesting we put it in backtics. Maybe using a cen makes sense too, but then it could be confused as itself being a Gall app.

  your agent will receive any updates the other agent sends out on that `path`.
  You can subscribe more than once to the same `path`, but each subscription
  must have a separate `wire` specified at the beginning of the [`%pass`
  card](#pass).
- `%watch-as`: This is the same as `%watch`, except Gall will convert updates to
  the given `mark` before delivering them to your agent.
- `%leave`: Unsubscribe. The subscription to cancel is determined by the `wire`
  at the beginning of the [`pass` card](#pass) rather than the subscription
  `path`, so its argument is just `~`.

**Examples**

```hoon
[%pass /some/wire %agent [~some-ship %some-agent] %watch /some/path]
[%pass /some/wire %agent [~some-ship %some-agent] %watch-as %some-mark /some/path]
[%pass /some/wire %agent [~some-ship %some-agent] %leave ~]
```

#### Pokes

Pokes are requests, actions, or just some data which you send to another agent.
Unlike subscriptions, these are just one-off messages.

A `%poke` contains a `cage` of some data. A `cage` is a cell of `[mark vase]`.
The `mark` is just a `@tas` like `%foo`, and corresponds to a mark file in the

> `/mar` directory. We'll cover marks in greater detail later. The `vase` contains

Reviewer: ```suggestion
`/mar` directory. We'll cover `mark`s in greater detail later. The `vase` contains
```

the actual data you're sending.

The `%poke-as` task is the same as `%poke` except Gall will convert the `cage`

> to the `mark` you specify before sending it off.

Reviewer: ```suggestion
The `%poke-as` task is the same as `%poke` except Gall will convert the `mark`
in the `cage` to the `mark` you specify before sending it off.
```

  ==
```

The letter at the beginning corresponds to the vane - `%b` for Behn, `%c` for
Clay, etc. After then vane letter comes the task. Each vane has an API with a
set of tasks that it will accept, and are defined in each vane's section of
`lull.hoon`. Each vane's tasks are documented on the API Reference page of its
section in the [Arvo documentation](/docs/arvo/arvo).

#### Examples

```hoon
[%pass /some/wire %arvo %b %wait (add ~m1 now.bowl)]
[%pass /some/wire %arvo %e %connect /some/url/path %some-agent]
```

## `%give`

```hoon
[%give gift]
```

The purpose of a `%give` card is to respond to a request made by another agent
or vane. More specifically, it's either for acknowledging a request, or for
sending out updates to subscribers. This is in contrast to a [`%pass`](#give)
card, which is essentially unsolicited.

A `%give` card contains a `gift:agent:gall` (henceforth just `gift`), which is
defined in `lull.hoon` as:

```hoon
+$  gift
  $%  [%fact paths=(list path) =cage]
      [%kick paths=(list path) ship=(unit ship)]
      [%watch-ack p=(unit tang)]
      [%poke-ack p=(unit tang)]
  ==
```

These can be divided into two categories:

### Acknowledgements

> `%watch-ack` and `%poke-ack` are sent in response to a `%watch` or `%poke`

Reviewer: Does this not also include `%watch-as` and `%poke-as`?
